lib/keyedqueue/api.go

```diff
@@ -0,0 +1,52 @@
+// Package keyedqueue provides a queue with keyed elements allowing elements
+// already on the queue to be easily replaced.
+package keyedqueue
+
+import (
+	"container/list"
+	"sync"
+)
+
+// Element represents a queue element.
+type Element interface {
+	// Key returns the key of this element. Returned key must support
+	// equality.
+	Key() interface{}
+}
+
+// Queue represents a queue of elements. Queue instances are safe to use
+// with multiple goroutines.
+type Queue struct {
+	lock       sync.Mutex
```

Comment:
Thanks for reminding me of RWMutex. I don't expect to do a lot more reading than writing. Adding to the queue is a mutating operation and so is removing from the queue. The only read-only op on a queue that comes to mind is getting the length, and I only do that when I expose proxima metrics.